`TestGatewayConformance/HTTPRouteReferenceGrant/Simple_HTTP_request_should_reach_web-backend` is flaky on CI

### Problem statement

`TestGatewayConformance/HTTPRouteReferenceGrant/Simple_HTTP_request_should_reach_web-backend` conformance test is flaky on CI:

https://github.com/Kong/kong-operator/actions/runs/19856530327/job/56895801259?pr=2792#step:7:2200

```
=== RUN   TestGatewayConformance/HTTPRouteReferenceGrant/Simple_HTTP_request_should_reach_web-backend
    httproute-reference-grant.go:54: 2025-12-02T11:20:42.147426707Z: Making GET request to host  via http://172.18.128.1/
    http.go:276: 2025-12-02T11:20:42.149142389Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 404. CRes: &{404 103 HTTP/1.1 map[Content-Length:[103] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:42 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[2ff8253d76e1a22cb053793186aea406] X-Kong-Response-Latency:[0]] <nil> []} (after 5.22µs)
    http.go:276: 2025-12-02T11:20:43.157308356Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 404. CRes: &{404 103 HTTP/1.1 map[Content-Length:[103] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:43 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[b0248b44a59887423b5fb4267390475a] X-Kong-Response-Latency:[0]] <nil> []} (after 1.002444317s)
    http.go:276: 2025-12-02T11:20:44.158800734Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 404. CRes: &{404 103 HTTP/1.1 map[Content-Length:[103] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:44 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[f533a962febefc4e30efee76d1f0753b] X-Kong-Response-Latency:[0]] <nil> []} (after 2.009998431s)
    http.go:276: 2025-12-02T11:20:45.160960318Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 404. CRes: &{404 103 HTTP/1.1 map[Content-Length:[103] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:45 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[9094c94e742aec5c90e6ca7c93a62f66] X-Kong-Response-Latency:[0]] <nil> []} (after 3.011875325s)
    http.go:276: 2025-12-02T11:20:46.16262604Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 404. CRes: &{404 103 HTTP/1.1 map[Content-Length:[103] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:46 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[4d6c047e7df4c16fea599b69dff06c76] X-Kong-Response-Latency:[0]] <nil> []} (after 4.013791487s)
    http.go:276: 2025-12-02T11:20:47.165001729Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 404. CRes: &{404 103 HTTP/1.1 map[Content-Length:[103] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:47 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[a7903343e408891e1bc331420bf0ee55] X-Kong-Response-Latency:[0]] <nil> []} (after 5.016255382s)
    http.go:276: 2025-12-02T11:20:48.166899689Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:48 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[a5b2d152b105aa7fea8aaa3323c35454] X-Kong-Response-Latency:[0]] <nil> []} (after 6.017957945s)
    http.go:276: 2025-12-02T11:20:49.169388862Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:49 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[90dfd33c62f28c2f39911d40bfd7f979] X-Kong-Response-Latency:[0]] <nil> []} (after 7.020036085s)
    http.go:276: 2025-12-02T11:20:50.171985003Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:50 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[d8d32f5ce9e4967145869e0088f8ba20] X-Kong-Response-Latency:[1]] <nil> []} (after 8.02238134s)
    http.go:276: 2025-12-02T11:20:51.174598725Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:51 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[8f7d1e3481e55227fb88027b84b1c716] X-Kong-Response-Latency:[1]] <nil> []} (after 9.025502644s)
    http.go:276: 2025-12-02T11:20:52.176716949Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:52 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[7fa44f762c78f17155a23d520f391945] X-Kong-Response-Latency:[0]] <nil> []} (after 10.027811842s)
    http.go:276: 2025-12-02T11:20:53.179051436Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:53 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[a0a2dd4ee588084fa3df482bed0ee24c] X-Kong-Response-Latency:[0]] <nil> []} (after 11.029458405s)
    http.go:276: 2025-12-02T11:20:54.181009676Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:54 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[637dca3e8d39e414ccff4e430a0efe01] X-Kong-Response-Latency:[0]] <nil> []} (after 12.032237659s)
    http.go:276: 2025-12-02T11:20:55.183468047Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:55 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[72426d33e7c2800515feec5717f53e06] X-Kong-Response-Latency:[0]] <nil> []} (after 13.034455965s)
    http.go:276: 2025-12-02T11:20:56.186021838Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:56 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[98c82a050089970b3ab8935a28a8fe41] X-Kong-Response-Latency:[0]] <nil> []} (after 14.037133819s)
    http.go:276: 2025-12-02T11:20:57.188428705Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:57 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[627b243b7349b11ae77ff83b3a4f5417] X-Kong-Response-Latency:[0]] <nil> []} (after 15.039266487s)
    http.go:276: 2025-12-02T11:20:58.193409067Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:58 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[5d2e480c2d06973cc2470f38ecab6640] X-Kong-Response-Latency:[1]] <nil> []} (after 16.042559708s)
    http.go:276: 2025-12-02T11:20:59.195505703Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:20:59 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[aeebe963bdd01133705d2e572d90395a] X-Kong-Response-Latency:[1]] <nil> []} (after 17.046744174s)
    http.go:276: 2025-12-02T11:21:00.197778565Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:21:00 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[e9480325c5ff6b293cc9521f8aebe073] X-Kong-Response-Latency:[0]] <nil> []} (after 18.049001498s)
    http.go:276: 2025-12-02T11:21:01.199650824Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:21:01 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[ab887478dfa89b4a77fac2400e063f37] X-Kong-Response-Latency:[1]] <nil> []} (after 19.05065745s)
    http.go:276: 2025-12-02T11:21:02.202129711Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:21:02 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[6de27f5f44fa110418143d39df24ed12] X-Kong-Response-Latency:[0]] <nil> []} (after 20.053216562s)
    http.go:276: 2025-12-02T11:21:03.204388701Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:21:03 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[2bbd92092e122a11f6192c19c485bd0e] X-Kong-Response-Latency:[0]] <nil> []} (after 21.055291283s)
    http.go:276: 2025-12-02T11:21:04.206879313Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:21:04 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[2e9bc11c31d2bf73ec2dbc509016e657] X-Kong-Response-Latency:[0]] <nil> []} (after 22.058049437s)
    http.go:276: 2025-12-02T11:21:05.208825585Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:21:05 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[a41b710ed4adb40cca653a6f014ab91a] X-Kong-Response-Latency:[0]] <nil> []} (after 23.060076147s)
    http.go:276: 2025-12-02T11:21:06.210878406Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:21:06 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[01b44b498cad4927edf51f5baf9f352d] X-Kong-Response-Latency:[0]] <nil> []} (after 24.062094897s)
    http.go:276: 2025-12-02T11:21:07.213035021Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:21:07 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[497e5b5a5d534e32dc2f6f0733c75e7f] X-Kong-Response-Latency:[0]] <nil> []} (after 25.064076825s)
    http.go:276: 2025-12-02T11:21:08.215746897Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:21:08 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[2353ba5b18f9c126bdb23f73cedb2e54] X-Kong-Response-Latency:[0]] <nil> []} (after 26.066719312s)
    http.go:276: 2025-12-02T11:21:09.217652735Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:21:09 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[f8a15d7134ad0234dce16d0b6584dd50] X-Kong-Response-Latency:[0]] <nil> []} (after 27.068934139s)
    http.go:276: 2025-12-02T11:21:10.219986894Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:21:10 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[78a71438f4f5aa4444a9ca63a464a741] X-Kong-Response-Latency:[0]] <nil> []} (after 28.071075727s)
    http.go:276: 2025-12-02T11:21:11.222945321Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Tue, 02 Dec 2025 11:21:11 GMT] Server:[kong/3.12.0.1-enterprise-edition] X-Kong-Request-Id:[2208d5553ac3f67c68ea645050a39672] X-Kong-Response-Latency:[0]] <nil> []} (after 29.073078115s)
    http.go:257: 2025-12-02T11:21:12.147961844Z: timeout while waiting after 30 attempts, 0/3 successes
--- FAIL: TestGatewayConformance/HTTPRouteReferenceGrant/Simple_HTTP_request_should_reach_web-backend (30.00s)
```

Still happening:

https://github.com/Kong/kong-operator/actions/runs/21671879173/job/62481655625#step:7:2100

```
...
    http.go:276: 2026-02-04T12:52:44.913877973Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Wed, 04 Feb 2026 12:52:44 GMT] Server:[kong/3.13.0.1-enterprise-edition] X-Kong-Request-Id:[0ac6f85a4b460f058c084522b50f5809] X-Kong-Response-Latency:[0]] <nil> []} (after 27.055100583s)
    http.go:276: 2026-02-04T12:52:45.915542841Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Wed, 04 Feb 2026 12:52:45 GMT] Server:[kong/3.13.0.1-enterprise-edition] X-Kong-Request-Id:[aeea197f69ba536aabd0d81adc14a06a] X-Kong-Response-Latency:[0]] <nil> []} (after 28.056789278s)
    http.go:276: 2026-02-04T12:52:46.917606513Z: Response expectation failed for request: {URL: {Scheme:http Opaque: User: Host:172.18.128.1 Path:/ RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}, Host: , Protocol: HTTP, Method: GET, Headers: map[X-Echo-Set-Header:[]], UnfollowRedirect: false, Server: , CertPem: <truncated>, KeyPem: <truncated>}  not ready yet: expected status code to be one of [200], got 503. CRes: &{503 113 HTTP/1.1 map[Content-Length:[113] Content-Type:[application/json; charset=utf-8] Date:[Wed, 04 Feb 2026 12:52:46 GMT] Server:[kong/3.13.0.1-enterprise-edition] X-Kong-Request-Id:[4d4ebf77bcbed948cde2ed1d5485a129] X-Kong-Response-Latency:[1]] <nil> []} (after 29.058627228s)
    http.go:257: 2026-02-04T12:52:47.858325111Z: timeout while waiting after 30 attempts, 0/3 successes
--- FAIL: TestGatewayConformance/HTTPRouteReferenceGrant/Simple_HTTP_request_should_reach_web-backend (30.00s)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -49,6 +49,20 @@
 - [v0.1.1](#v011)
 - [v0.1.0](#v010)
 
+## Unreleased
+
+### Fixes
+
+- Hybridgateway: release Gateway API route finalizers once generated Kong
+  resource delete requests have been issued, so immediate same-name route
+  re-creates are not blocked by child resource finalizers.
+  [#4465](https://github.com/Kong/kong-operator/pull/4465)
+- Hybridgateway: use route-scoped `KongService` names for `HTTPRoute` rules
+  whose backendRefs resolve to no valid targets. This avoids Konnect name
+  conflicts with valid backend services while keeping normally generated service
+  names unchanged.
+  [#4437](https://github.com/Kong/kong-operator/pull/4437)
+
 ## [v2.2.0]
 
 > Release date: 2026-06-05
@@ -253,15 +267,6 @@
 
 ### Fixes
 
-- Hybridgateway: release Gateway API route finalizers once generated Kong
-  resource delete requests have been issued, so immediate same-name route
-  re-creates are not blocked by child resource finalizers.
-  [#4465](https://github.com/Kong/kong-operator/pull/4465)
-- Hybridgateway: use route-scoped `KongService` names for `HTTPRoute` rules
-  whose backendRefs resolve to no valid targets. This avoids Konnect name
-  conflicts with valid backend services while keeping normally generated service
-  names unchanged.
-  [#4437](https://github.com/Kong/kong-operator/pull/4437)
 - Hybridgateway: fix `KongRoute` name collisions when an `HTTPRoute` attaches
   to multiple listener-scoped `ParentRef`s on the same `Gateway`, enabling the
   `HTTPRouteListenerHostnameMatching` Gateway API conformance test for Hybrid

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -257,6 +257,11 @@
   resource delete requests have been issued, so immediate same-name route
   re-creates are not blocked by child resource finalizers.
   [#4465](https://github.com/Kong/kong-operator/pull/4465)
+- Hybridgateway: use route-scoped `KongService` names for `HTTPRoute` rules
+  whose backendRefs resolve to no valid targets. This avoids Konnect name
+  conflicts with valid backend services while keeping normally generated service
+  names unchanged.
+  [#4437](https://github.com/Kong/kong-operator/pull/4437)
 - Hybridgateway: fix `KongRoute` name collisions when an `HTTPRoute` attaches
   to multiple listener-scoped `ParentRef`s on the same `Gateway`, enabling the
   `HTTPRouteListenerHostnameMatching` Gateway API conformance test for Hybrid

--- a/controller/hybridgateway/converter/http_route.go
+++ b/controller/hybridgateway/converter/http_route.go
@@ -408,8 +408,36 @@ func (c *httpRouteConverter) translate(ctx context.Context, logger logr.Logger) 
 
 			upstreamName := namegen.NewKongUpstreamNameForHTTPRouteRule(c.route, cp, rule)
 
+			// Build the KongTarget resources before the service so fallback services for
+			// invalid backends can use a route-scoped name and avoid colliding with normal backend services.
+			targets, err := target.TargetsForBackendRefs(
+				ctx,
+				logger.WithValues("upstream", upstreamName),
+				c.Client,
+				c.route,
+				rule.BackendRefs,
+				&pRef,
+				upstreamName,
+				c.fqdnMode,
+				c.clusterDomain,
+			)
+			if err != nil {
+				log.Error(logger, err, "Failed to translate KongTarget resources for rule, skipping rule",
+					"upstream", upstreamName,
+					"backendRefs", rule.BackendRefs,
+					"parentRef", pRef)
+				translationErrors = append(translationErrors, fmt.Errorf("failed to translate KongTarget resources for upstream %s: %w", upstreamName, err))
+				continue
+			}
+			log.Debug(logger, "Successfully translated KongTarget resources", "upstream", upstreamName, "targetCount", len(targets))
+
+			serviceNameOverride := ""
+			if len(rule.BackendRefs) > 0 && len(targets) == 0 {
+				serviceNameOverride = namegen.NewKongServiceNameForHTTPRouteRuleBackendNotFound(c.route, cp, rule)
+			}
+
 			// Build the KongService resource (and optionally a KongCertificate + KongReferenceGrant for client-cert).
-			servicePtr, certPtr, grantPtr, err := service.ServiceForRule(ctx, logger, c.Client, c.route, rule, &pRef, cp, upstreamName)
+			servicePtr, certPtr, grantPtr, err := service.ServiceForRuleWithName(ctx, logger, c.Client, c.route, rule, &pRef, cp, upstreamName, serviceNameOverride)
 			if err != nil {
 				log.Error(logger, err, "Failed to translate KongService resource, skipping rule",
 					"controlPlane", cp.KonnectNamespacedRef,
@@ -479,29 +507,6 @@ func (c *httpRouteConverter) translate(ctx context.Context, logger logr.Logger) 
 					}
 				}
 			}
-
-			// Build the KongTarget resources.
-			// Leave them as the last step since we want everything fully configured before enabling the traffic to the backends.
-			targets, err := target.TargetsForBackendRefs(
-				ctx,
-				logger.WithValues("upstream", upstreamName),
-				c.Client,
-				c.route,
-				rule.BackendRefs,
-				&pRef,
-				upstreamName,
-				c.fqdnMode,
-				c.clusterDomain,
-			)
-			if err != nil {
-				log.Error(logger, err, "Failed to translate KongTarget resources for rule, skipping rule",
-					"upstream", upstreamName,
-					"backendRefs", rule.BackendRefs,
-					"parentRef", pRef)
-				translationErrors = append(translationErrors, fmt.Errorf("failed to translate KongTarget resources for upstream %s: %w", upstreamName, err))
-				continue
-			}
-			log.Debug(logger, "Successfully translated KongTarget resources", "upstream", upstreamName, "targetCount", len(targets))
 
 			ruleOutputs := make([]client.Object, 0, 1+len(routes)+len(filterOutputs)+len(targets)+4)
 			if len(rule.BackendRefs) == 0 || len(targets) > 0 {

--- a/controller/hybridgateway/converter/http_route.go
+++ b/controller/hybridgateway/converter/http_route.go
@@ -18,6 +18,7 @@ import (
 	hybridgatewayerrors "github.com/kong/kong-operator/v2/controller/hybridgateway/errors"
 	"github.com/kong/kong-operator/v2/controller/hybridgateway/kongroute"
 	"github.com/kong/kong-operator/v2/controller/hybridgateway/metadata"
+	"github.com/kong/kong-operator/v2/controller/hybridgateway/namegen"
 	"github.com/kong/kong-operator/v2/controller/hybridgateway/plugin"
 	"github.com/kong/kong-operator/v2/controller/hybridgateway/pluginbinding"
 	"github.com/kong/kong-operator/v2/controller/hybridgateway/refs"
@@ -405,18 +406,7 @@ func (c *httpRouteConverter) translate(ctx context.Context, logger logr.Logger) 
 				"matchCount", len(rule.Matches),
 				"filterCount", len(rule.Filters))
 
-			// Build the KongUpstream resource.
-			upstreamPtr, err := upstream.UpstreamForRule(ctx, logger, c.Client, c.route, rule, &pRef, cp)
-			if err != nil {
-				log.Error(logger, err, "Failed to translate KongUpstream resource for rule, skipping rule",
-					"controlPlane", cp.KonnectNamespacedRef)
-				translationErrors = append(translationErrors, fmt.Errorf("failed to translate KongUpstream resource: %w", err))
-				continue
-			}
-			upstreamName := upstreamPtr.Name
-			c.outputStore = append(c.outputStore, upstreamPtr)
-			log.Debug(logger, "Successfully translated KongUpstream resource",
-				"upstream", upstreamName)
+			upstreamName := namegen.NewKongUpstreamNameForHTTPRouteRule(c.route, cp, rule)
 
 			// Build the KongService resource (and optionally a KongCertificate + KongReferenceGrant for client-cert).
 			servicePtr, certPtr, grantPtr, err := service.ServiceForRule(ctx, logger, c.Client, c.route, rule, &pRef, cp, upstreamName)
@@ -427,19 +417,7 @@ func (c *httpRouteConverter) translate(ctx context.Context, logger logr.Logger) 
 				translationErrors = append(translationErrors, fmt.Errorf("failed to translate KongService for rule: %w", err))
 				continue
 			}
-			// Append KongReferenceGrant before KongCertificate so the grant exists when the cert is applied.
-			if grantPtr != nil {
-				c.outputStore = append(c.outputStore, grantPtr)
-				log.Debug(logger, "Successfully translated KongReferenceGrant resource", "grant", grantPtr.Name)
-			}
-			if certPtr != nil {
-				c.outputStore = append(c.outputStore, certPtr)
-				log.Debug(logger, "Successfully translated KongCertificate resource", "cert", certPtr.Name)
-			}
 			serviceName := servicePtr.Name
-			c.outputStore = append(c.outputStore, servicePtr)
-			log.Debug(logger, "Successfully translated KongService resource",
-				"service", serviceName)
 
 			// Build one KongRoute per match in the rule.
 			// Gateway API semantics require OR across matches within a rule
@@ -454,16 +432,10 @@ func (c *httpRouteConverter) translate(ctx context.Context, logger logr.Logger) 
 				translationErrors = append(translationErrors, fmt.Errorf("failed to translate KongRoutes for rule %d: %w", ruleIndex, err))
 				continue
 			}
-			for _, r := range routes {
-				routeName := r.Name
-				c.outputStore = append(c.outputStore, r)
-				log.Debug(logger, "Successfully translated KongRoute resource",
-					"route", routeName)
-			}
-
 			// Build the KongPlugin and KongPluginBinding resources.
 			log.Debug(logger, "Processing filters for rule",
 				"filterCount", len(rule.Filters))
+			filterOutputs := make([]client.Object, 0)
 
 			for _, filter := range rule.Filters {
 				plugins, err := plugin.PluginsForFilter(ctx, logger, c.Client, c.route, rule, filter, &pRef)
@@ -474,9 +446,10 @@ func (c *httpRouteConverter) translate(ctx context.Context, logger logr.Logger) 
 					continue
 				}
 
-				for _, plugin := range plugins {
-					pluginName := plugin.Name
-					c.outputStore = append(c.outputStore, &plugin)
+				for i := range plugins {
+					pluginObj := &plugins[i]
+					pluginName := pluginObj.Name
+					filterOutputs = append(filterOutputs, pluginObj)
 					// Create a KongPluginBinding to bind the KongPlugin to each KongRoute generated for the rule.
 					for _, r := range routes {
 						bindingPtr, err := pluginbinding.BindingForPluginAndRoute(
@@ -497,7 +470,7 @@ func (c *httpRouteConverter) translate(ctx context.Context, logger logr.Logger) 
 							continue
 						}
 						bindingName := bindingPtr.Name
-						c.outputStore = append(c.outputStore, bindingPtr)
+						filterOutputs = append(filterOutputs, bindingPtr)
 
 						log.Debug(logger, "Successfully translated KongPlugin and KongPluginBinding resources",
 							"plugin", pluginName,
@@ -528,12 +501,42 @@ func (c *httpRouteConverter) translate(ctx context.Context, logger logr.Logger) 
 				translationErrors = append(translationErrors, fmt.Errorf("failed to translate KongTarget resources for upstream %s: %w", upstreamName, err))
 				continue
 			}
-			log.Debug(logger, "Successfully translated KongTarget resources",
-				"upstream", upstreamName,
-				"targetCount", len(targets))
-			for _, tgt := range targets {
-				c.outputStore = append(c.outputStore, &tgt)
+			log.Debug(logger, "Successfully translated KongTarget resources", "upstream", upstreamName, "targetCount", len(targets))
+
+			ruleOutputs := make([]client.Object, 0, 1+len(routes)+len(filterOutputs)+len(targets)+4)
+			if len(rule.BackendRefs) == 0 || len(targets) > 0 {
+				upstreamPtr, err := upstream.UpstreamForRule(ctx, logger, c.Client, c.route, rule, &pRef, cp)
+				if err != nil {
+					log.Error(logger, err, "Failed to translate KongUpstream resource for rule, skipping rule",
+						"controlPlane", cp.KonnectNamespacedRef)
+					translationErrors = append(translationErrors, fmt.Errorf("failed to translate KongUpstream resource: %w", err))
+					continue
+				}
+				ruleOutputs = append(ruleOutputs, upstreamPtr)
+				log.Debug(logger, "Successfully translated KongUpstream resource", "upstream", upstreamName)
+			} else {
+				log.Debug(logger, "Skipping KongUpstream translation because no valid backend targets were produced",
+					"upstream", upstreamName,
+					"backendRefCount", len(rule.BackendRefs))
 			}
+
+			// Append KongReferenceGrant before KongCertificate so the grant exists when the cert is applied.
+			if grantPtr != nil {
+				ruleOutputs = append(ruleOutputs, grantPtr)
+				log.Debug(logger, "Successfully translated KongReferenceGrant resource", "grant", grantPtr.Name)
+			}
+			if certPtr != nil {
+				ruleOutputs = append(ruleOutputs, certPtr)
+				log.Debug(logger, "Successfully translated KongCertificate resource", "cert", certPtr.Name)
+			}
+			ruleOutputs = append(ruleOutputs, servicePtr)
+			log.Debug(logger, "Successfully translated KongService resource", "service", serviceName)
+			for _, r := range routes {
+				routeName := r.Name
+				ruleOutputs = append(ruleOutputs, r)
+				log.Debug(logger, "Successfully translated KongRoute resource", "route", routeName)
+			}
+			ruleOutputs = append(ruleOutputs, filterOutputs...)
 
 			if len(rule.BackendRefs) > 0 && len(targets) == 0 {
 				terminationPlugin, err := plugin.RequestTerminationForBackendNotFound(
@@ -551,7 +554,6 @@ func (c *httpRouteConverter) translate(ctx context.Context, logger logr.Logger) 
 					translationErrors = append(translationErrors, fmt.Errorf("failed to translate request-termination plugin for service %s: %w", serviceName, err))
 					continue
 				}
-				c.outputStore = append(c.outputStore, terminationPlugin)
 
 				bindingPtr, err := pluginbinding.BindingForPluginAndService(
 					ctx,
@@ -570,8 +572,14 @@ func (c *httpRouteConverter) translate(ctx context.Context, logger logr.Logger) 
 					translationErrors = append(translationErrors, fmt.Errorf("failed to bind request-termination plugin %s to service %s: %w", terminationPlugin.Name, serviceName, err))
 					continue
 				}
-				c.outputStore = append(c.outputStore, bindingPtr)
+				ruleOutputs = append(ruleOutputs, terminationPlugin, bindingPtr)
 			}
+
+			for i := range targets {
+				ruleOutputs = append(ruleOutputs, &targets[i])
+			}
+
+			c.outputStore = append(c.outputStore, ruleOutputs...)
 		}
 	}
 

--- a/controller/hybridgateway/converter/http_route.go
+++ b/controller/hybridgateway/converter/http_route.go
@@ -508,7 +508,7 @@ func (c *httpRouteConverter) translate(ctx context.Context, logger logr.Logger) 
 				}
 			}
 
-			ruleOutputs := make([]client.Object, 0, 1+len(routes)+len(filterOutputs)+len(targets)+4)
+			var ruleOutputs []client.Object
 			if len(rule.BackendRefs) == 0 || len(targets) > 0 {
 				upstreamPtr, err := upstream.UpstreamForRule(ctx, logger, c.Client, c.route, rule, &pRef, cp)
 				if err != nil {

--- a/controller/hybridgateway/converter/http_route_converter_test.go
+++ b/controller/hybridgateway/converter/http_route_converter_test.go
@@ -24,6 +24,7 @@ import (
 
 	configurationv1 "github.com/kong/kong-operator/v2/api/configuration/v1"
 	configurationv1alpha1 "github.com/kong/kong-operator/v2/api/configuration/v1alpha1"
+	"github.com/kong/kong-operator/v2/controller/hybridgateway/namegen"
 	gwtypes "github.com/kong/kong-operator/v2/internal/types"
 	"github.com/kong/kong-operator/v2/modules/manager/scheme"
 	"github.com/kong/kong-operator/v2/pkg/consts"
@@ -352,15 +353,15 @@ func TestHTTPRouteConverter_Translate(t *testing.T) {
 				t.Helper()
 
 				var (
-					serviceName string
-					pluginObj   *configurationv1.KongPlugin
-					bindingObj  *configurationv1alpha1.KongPluginBinding
+					serviceObj *configurationv1alpha1.KongService
+					pluginObj  *configurationv1.KongPlugin
+					bindingObj *configurationv1alpha1.KongPluginBinding
 				)
 
 				for _, obj := range store {
 					switch typed := obj.(type) {
 					case *configurationv1alpha1.KongService:
-						serviceName = typed.Name
+						serviceObj = typed
 					case *configurationv1.KongPlugin:
 						pluginObj = typed
 					case *configurationv1alpha1.KongPluginBinding:
@@ -368,7 +369,14 @@ func TestHTTPRouteConverter_Translate(t *testing.T) {
 					}
 				}
 
+				require.NotNil(t, serviceObj)
+				serviceName := serviceObj.Name
+				normalRoute := newHTTPRouteForTranslation([]string{"api.example.com"}, []gwtypes.HTTPBackendRef{
+					newBackendRef("backend"),
+				}, nil)
+				normalServiceName := namegen.NewKongServiceNameForHTTPRouteRule(normalRoute, serviceObj.Spec.ControlPlaneRef, normalRoute.Spec.Rules[0])
 				require.NotEmpty(t, serviceName)
+				assert.NotEqual(t, normalServiceName, serviceName)
 				require.NotNil(t, pluginObj)
 				require.NotNil(t, bindingObj)
 				assert.Equal(t, "request-termination", pluginObj.PluginName)

--- a/controller/hybridgateway/converter/http_route_converter_test.go
+++ b/controller/hybridgateway/converter/http_route_converter_test.go
@@ -338,16 +338,16 @@ func TestHTTPRouteConverter_Translate(t *testing.T) {
 				fakeClient := fake.NewClientBuilder().WithScheme(scheme.Get()).WithObjects(objects...).Build()
 				return newHTTPRouteConverter(route, fakeClient, false, "").(*httpRouteConverter)
 			},
-			wantCount: 5,
+			wantCount: 4,
 			wantOutputs: outputCount{
-				upstreams: 1,
+				upstreams: 0,
 				services:  1,
 				routes:    1,
 				targets:   0,
 				bindings:  1,
 				plugins:   1,
 			},
-			wantStoreLen: 5,
+			wantStoreLen: 4,
 			assertFn: func(t *testing.T, store []client.Object) {
 				t.Helper()
 
@@ -397,16 +397,16 @@ func TestHTTPRouteConverter_Translate(t *testing.T) {
 				fakeClient := fake.NewClientBuilder().WithScheme(scheme.Get()).WithObjects(objects...).Build()
 				return newHTTPRouteConverter(route, fakeClient, false, "").(*httpRouteConverter)
 			},
-			wantCount: 5,
+			wantCount: 4,
 			wantOutputs: outputCount{
-				upstreams: 1,
+				upstreams: 0,
 				services:  1,
 				routes:    1,
 				targets:   0,
 				bindings:  1,
 				plugins:   1,
 			},
-			wantStoreLen: 5,
+			wantStoreLen: 4,
 			assertFn: func(t *testing.T, store []client.Object) {
 				t.Helper()
 
@@ -545,16 +545,16 @@ func TestHTTPRouteConverter_Translate(t *testing.T) {
 				fakeClient := fake.NewClientBuilder().WithScheme(scheme.Get()).WithObjects(objects...).Build()
 				return newHTTPRouteConverter(route, fakeClient, false, "").(*httpRouteConverter)
 			},
-			wantCount: 9,
+			wantCount: 8,
 			wantOutputs: outputCount{
-				upstreams: 2,
+				upstreams: 1,
 				services:  2,
 				routes:    2,
 				targets:   1,
 				bindings:  1,
 				plugins:   1,
 			},
-			wantStoreLen: 9,
+			wantStoreLen: 8,
 			assertFn: func(t *testing.T, store []client.Object) {
 				t.Helper()
 
@@ -603,16 +603,16 @@ func TestHTTPRouteConverter_Translate(t *testing.T) {
 				fakeClient := fake.NewClientBuilder().WithScheme(scheme.Get()).WithObjects(objects...).Build()
 				return newHTTPRouteConverter(route, fakeClient, false, "").(*httpRouteConverter)
 			},
-			wantCount: 5,
+			wantCount: 4,
 			wantOutputs: outputCount{
-				upstreams: 1,
+				upstreams: 0,
 				services:  1,
 				routes:    1,
 				targets:   0,
 				bindings:  1,
 				plugins:   1,
 			},
-			wantStoreLen: 5,
+			wantStoreLen: 4,
 			assertFn: func(t *testing.T, store []client.Object) {
 				t.Helper()
 
@@ -980,12 +980,9 @@ func TestHTTPRouteConverter_Translate(t *testing.T) {
 					Build()
 				return newHTTPRouteConverter(route, fakeClient, false, "").(*httpRouteConverter)
 			},
-			wantErr:    true,
-			wantErrSub: "failed to translate KongService for rule",
-			wantOutputs: outputCount{
-				upstreams: 1,
-			},
-			wantStoreLen: 1,
+			wantErr:      true,
+			wantErrSub:   "failed to translate KongService for rule",
+			wantStoreLen: 0,
 		},
 		{
 			name: "returns error when route translation fails",
@@ -1006,13 +1003,9 @@ func TestHTTPRouteConverter_Translate(t *testing.T) {
 					Build()
 				return newHTTPRouteConverter(route, fakeClient, false, "").(*httpRouteConverter)
 			},
-			wantErr:    true,
-			wantErrSub: "failed to translate KongRoutes for rule",
-			wantOutputs: outputCount{
-				upstreams: 1,
-				services:  1,
-			},
-			wantStoreLen: 2,
+			wantErr:      true,
+			wantErrSub:   "failed to translate KongRoutes for rule",
+			wantStoreLen: 0,
 		},
 		{
 			name: "returns error when plugin binding fails",
@@ -1068,14 +1061,9 @@ func TestHTTPRouteConverter_Translate(t *testing.T) {
 					Build()
 				return newHTTPRouteConverter(route, fakeClient, false, "").(*httpRouteConverter)
 			},
-			wantErr:    true,
-			wantErrSub: "failed to translate KongTarget resources for upstream",
-			wantOutputs: outputCount{
-				upstreams: 1,
-				services:  1,
-				routes:    1,
-			},
-			wantStoreLen: 3,
+			wantErr:      true,
+			wantErrSub:   "failed to translate KongTarget resources for upstream",
+			wantStoreLen: 0,
 		},
 	}
 

--- a/controller/hybridgateway/namegen/name.go
+++ b/controller/hybridgateway/namegen/name.go
@@ -24,6 +24,10 @@ const (
 	// parentRefPrefix is the prefix used when including a ParentRef identifier.
 	parentRefPrefix = "pr"
 
+	// backendNotFoundPrefix is the prefix used for service names that intentionally
+	// route to request-termination because no backend target could be produced.
+	backendNotFoundPrefix = "bnf"
+
 	// namegenPrefix is used as the prefix for hashed names when concatenated
 	// components exceed the maximum Kubernetes resource name length.
 	namegenPrefix = "ngn"
@@ -109,6 +113,32 @@ func NewKongServiceNameForHTTPRouteRule(route *gwtypes.HTTPRoute, cp *commonv1al
 	return newNameWithHashSuffix(readableElements, hashElements)
 }
 
+// NewKongServiceNameForHTTPRouteRuleBackendNotFound generates a route-scoped KongService name
+// for rules that have BackendRefs but no valid backend targets. These services get a
+// request-termination plugin, so they must not share the normal backend service name.
+func NewKongServiceNameForHTTPRouteRuleBackendNotFound(
+	route *gwtypes.HTTPRoute,
+	cp *commonv1alpha1.ControlPlaneRef,
+	rule gatewayv1.HTTPRouteRule,
+) string {
+	readableElements := append(
+		[]string{httpProcolPrefix},
+		backendRefDisplayNames(route.Namespace, rule.BackendRefs)...,
+	)
+	hashElements := []string{
+		defaultCPPrefix + utils.Hash32(cp),
+		backendNotFoundPrefix + utils.Hash32(struct {
+			Namespace string
+			Name      string
+		}{
+			Namespace: route.Namespace,
+			Name:      route.Name,
+		}),
+		hashForHTTPRouteRuleServiceLikeName(route, rule),
+	}
+	return newNameWithHashSuffix(readableElements, hashElements)
+}
+
 // NewKongServiceNameForTLSRouteRule generates a KongService name based on the ControlPlaneRef and TLSRouteRule passed as arguments.
 func NewKongServiceNameForTLSRouteRule(route *gwtypes.TLSRoute, cp *commonv1alpha1.ControlPlaneRef, rule gatewayv1.TLSRouteRule) string {
 	readableElements := append(
@@ -124,6 +154,14 @@ func hashElementsForServiceLikeName(
 	cp *commonv1alpha1.ControlPlaneRef,
 	rule gatewayv1.HTTPRouteRule,
 ) []string {
+	hash := hashForHTTPRouteRuleServiceLikeName(route, rule)
+	return []string{
+		defaultCPPrefix + utils.Hash32(cp),
+		hash,
+	}
+}
+
+func hashForHTTPRouteRuleServiceLikeName(route *gwtypes.HTTPRoute, rule gatewayv1.HTTPRouteRule) string {
 	var hash string
 	if len(rule.BackendRefs) > 0 {
 		hash = utils.Hash32(rule.BackendRefs)
@@ -142,10 +180,7 @@ func hashElementsForServiceLikeName(
 		hash = utils.Hash32(zeroBackendRuleIdentity)
 	}
 
-	return []string{
-		defaultCPPrefix + utils.Hash32(cp),
-		hash,
-	}
+	return hash
 }
 
 func hashElementsForServiceLikeNameTLSRouteRule(

--- a/controller/hybridgateway/namegen/name_test.go
+++ b/controller/hybridgateway/namegen/name_test.go
@@ -276,6 +276,27 @@ func TestNewKongServiceName(t *testing.T) {
 	}
 }
 
+func TestNewKongServiceNameForHTTPRouteRuleBackendNotFound(t *testing.T) {
+	backendNS := gatewayv1.Namespace("gateway-conformance-web-backend")
+	port := gatewayv1.PortNumber(8080)
+	rule := gatewayv1.HTTPRouteRule{
+		BackendRefs: []gatewayv1.HTTPBackendRef{
+			testBackendRef("web-backend", &backendNS, &port),
+		},
+	}
+	routeA := testRoute("gateway-conformance-infra", "invalid-cross-namespace-backend-ref")
+	routeB := testRoute("gateway-conformance-infra", "reference-grant")
+	cp := testControlPlaneRef("same-namespace")
+
+	normalName := NewKongServiceNameForHTTPRouteRule(routeA, cp, rule)
+	fallbackNameA := NewKongServiceNameForHTTPRouteRuleBackendNotFound(routeA, cp, rule)
+	fallbackNameB := NewKongServiceNameForHTTPRouteRuleBackendNotFound(routeB, cp, rule)
+
+	assert.NotEqual(t, normalName, fallbackNameA)
+	assert.NotEqual(t, fallbackNameA, fallbackNameB)
+	assert.Contains(t, fallbackNameA, backendNotFoundPrefix)
+}
+
 func TestNewKongServiceName_Equality(t *testing.T) {
 	tests := []struct {
 		name  string

--- a/controller/hybridgateway/service/service.go
+++ b/controller/hybridgateway/service/service.go
@@ -60,7 +60,25 @@ func ServiceForRule[
 	cp *commonv1alpha1.ControlPlaneRef,
 	upstreamName string,
 ) (kongService *configurationv1alpha1.KongService, kongCertificate *configurationv1alpha1.KongCertificate, kongReferenceGrant *configurationv1alpha1.KongReferenceGrant, err error) {
+	return ServiceForRuleWithName(ctx, logger, cl, parentRoute, rule, pRef, cp, upstreamName, "")
+}
 
+// ServiceForRuleWithName behaves like ServiceForRule but uses serviceName when it is not empty.
+func ServiceForRuleWithName[
+	T gwtypes.SupportedRoute,
+	TPtr gwtypes.SupportedRoutePtr[T],
+	R gwtypes.SupportedRouteRule,
+](
+	ctx context.Context,
+	logger logr.Logger,
+	cl client.Client,
+	parentRoute TPtr,
+	rule R,
+	pRef *gwtypes.ParentReference,
+	cp *commonv1alpha1.ControlPlaneRef,
+	upstreamName string,
+	serviceNameOverride string,
+) (kongService *configurationv1alpha1.KongService, kongCertificate *configurationv1alpha1.KongCertificate, kongReferenceGrant *configurationv1alpha1.KongReferenceGrant, err error) {
 	var serviceName string
 	var namespace string
 	var backendRefs []gwtypes.BackendRef
@@ -92,6 +110,10 @@ func ServiceForRule[
 	// Should be unreachable.
 	default:
 		return nil, nil, nil, fmt.Errorf("failed to build KongService: unsupported route type: %T", parentRoute)
+	}
+
+	if serviceNameOverride != "" {
+		serviceName = serviceNameOverride
 	}
 
 	// Resolve service attributes once, outside the switch — future route types only add a case above.

--- a/test/conformance/conformance_test.go
+++ b/test/conformance/conformance_test.go
@@ -125,6 +125,9 @@ func runConformance(
 	timeoutConfig.GatewayStatusMustHaveListeners = conformanceLooserTimeout
 	timeoutConfig.GatewayListenersMustHaveConditions = conformanceLooserTimeout
 	timeoutConfig.HTTPRouteMustHaveCondition = conformanceLooserTimeout
+	if gwType == hybridGateway {
+		timeoutConfig.MaxTimeToConsistency = conformanceLooserTimeout
+	}
 
 	opts := conformance.DefaultOptions(t)
 	// It takes default conformance suite configuration manifests from provided location.

--- a/test/conformance/skipped_tests_test.go
+++ b/test/conformance/skipped_tests_test.go
@@ -38,10 +38,6 @@ var skippedTestsForHybrid = []string{
 	// Extended profile.
 	tests.HTTPRouteRewriteHost.ShortName,
 	tests.HTTPRouteRewritePath.ShortName,
-
-	// This test is skipped because it proven to be flaky.
-	// TODO: https://github.com/Kong/kong-operator/issues/2793
-	tests.HTTPRouteReferenceGrant.ShortName,
 }
 
 // skippedTestsForConfig returns the list of skipped tests for the given router flavor and gateway type.

